lib/phoenix/router/helpers.ex

```diff
@@ -176,20 +176,37 @@ defmodule Phoenix.Router.Helpers do
       defp to_param(true), do: "true"
       defp to_param(data), do: Phoenix.Param.to_param(data)
 
-      defp segments(segments, [], _reserved) do
+      defp segments!(segments, [], _reserved, _opts) do
         segments
       end
 
-      defp segments(segments, query, reserved) do
+      defp segments!(segments, query, reserved, _opts) when is_list(query) or is_map(query) do
         dict = for {k, v} <- query,
                not ((k = to_string(k)) in reserved),
                do: {k, v}
 
+
         case Conn.Query.encode dict, &to_param/1 do
           "" -> segments
           o  -> segments <> "?" <> o
         end
       end
+
+      defp segments!(segments, query, reserved, {helper, opts, var_length}) do
+        arity = var_length + 3
+        call_vars = if var_length > 0, do: Enum.map(1..(var_length), &(":arg#{&1}")), else: []
+
+        raise ArgumentError, """
+        #{__MODULE__}.#{helper}_path/#{arity} called with invalid params. The last argument to \
+        this function should be a keyword list or a map.
+        For example:
+
+          Router.#{helper}_path(conn, :#{opts}, #{Enum.join(call_vars ++ [""], ", ")}page: 5, per_page: 10)
```

Comment:
If `call_vars` is empty, won't we have two commas? Maybe you want to use intersperse instead?

```iex
iex(2)> "foo, bar, #{Enum.intersperse ~w(arg1 arg2 arg3), ", "}, done"
"foo, bar, arg1, arg2, arg3, done"
```

Reply:
Intersperse in this case results in two commas. That's why I append an empty element - going to change it to a prepend though..

```elixir
"foo, bar, #{Enum.intersperse ~w(), ", "}, done"              
"foo, bar, , done"
```

Reply:
@Gazler I am silly. The interperse would be the same as join. I meant to say a map:

```iex
iex(3)> "foo, bar, #{Enum.map ~w(arg1 arg2 arg3), &[&1 | ", "]}done"
"foo, bar, arg1, arg2, arg3, done"
iex(4)> "foo, bar, #{Enum.map ~w(), &[&1 | ", "]}done"
"foo, bar, done"
```

But that is probably not better or worse than what we have.